File: src/adapters/dockle.ts
import type { NormalizedCompliance } from '../types/index.js';
import { mapDockleCategory, mapDockleSeverity } from './severity.js';

export function extractDockleCompliance(report: any): NormalizedCompliance[] {
  const findings: NormalizedCompliance[] = [];

  if (report?.details) {
    for (const detail of report.details) {
      for (const alert of detail.alerts || []) {
        findings.push({
          source: 'dockle',
          ruleId: detail.code,
          ruleName: detail.title,
          category: mapDockleCategory(detail.level),
          severity: mapDockleSeverity(detail.level),
          message: typeof alert === 'string' ? alert : alert?.message ?? String(alert),
        });
      }
    }
  }

  return findings;
}

File: internal/adapter/dockle.go
package adapter

import (
	"fmt"

	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

// ExtractDockleCompliance extracts normalized compliance findings from dockle JSON.
func ExtractDockleCompliance(report interface{}) []types.NormalizedCompliance {
	var findings []types.NormalizedCompliance

	data, ok := report.(map[string]interface{})
	if !ok {
		return findings
	}

	details, ok := data["details"].([]interface{})
	if !ok {
		return findings
	}

	for _, d := range details {
		detail, ok := d.(map[string]interface{})
		if !ok {
			continue
		}

		alerts, ok := detail["alerts"].([]interface{})
		if !ok {
			continue
		}

		level := getString(detail, "level")

		for _, a := range alerts {
			var message string
			switch v := a.(type) {
			case string:
				message = v
			case map[string]interface{}:
				message = getString(v, "message")
				if message == "" {
					message = fmt.Sprintf("%v", v)
				}
			default:
				message = fmt.Sprintf("%v", v)
			}

			findings = append(findings, types.NormalizedCompliance{
				Source:   "dockle",
				RuleID:   getString(detail, "code"),
				RuleName: getString(detail, "title"),
				Category: MapDockleCategory(level),
				Severity: MapDockleSeverity(level),
				Message:  message,
			})
		}
	}

	return findings
}
